### ISSUE TITLE
Add rule to detect and forbid `COPY . .` usage

### ISSUE THREAD

**Issue by scraly** [hadolint/hadolint#1199]
Hi,

As the software supply chain attacks are growing dangerously, I would like to propose a new linting rule that detects and optionally forbids the usage of:

    COPY . .

Why?

Using `COPY . .` in Dockerfiles is considered a bad practice for several reasons:

1. **Build context leakage**
   It copies the entire build context into the image, including potentially sensitive or unnecessary files such as:
   - `.git/`
   - local configuration files
   - `.env` files
   - caches or temporary files

Indeed, a Dockerfile with this instruction can copies .git directory and then deletes it, but the .git content remains in the previous layer and can be extracted! So we need to be alerted.

2. **Non-reproducible builds**
   It makes it harder to understand exactly what is included in the image.

3. **Security concerns**
   Developers may accidentally include secrets or private files in the image without realizing it.

4. **Best practice recommendation**
   Most production Dockerfiles prefer explicit file copying, e.g.:

       COPY go.mod go.sum ./
       COPY internal/ ./internal

### Suggested rule behavior

- Detect `COPY . .`
- Optionally provide a warning or error (configurable severity)
- Suggest replacing with explicit file/directory copies

### Example message

    Avoid using `COPY . .` as it copies the entire build context.
    Prefer explicit COPY instructions to improve security and build reproducibility.

I know the existence of the rule `DL3045` but it dones not really alert/deny about the usage of "COPY . .".

Thx :)

**Comment by m-ildefons** [comment]
Hi Aurelie,

thanks for the suggestion.

I think this rule may be a bit too prone to false-positives in many different situations:

1) When users are utilizing [.dockerignore files](https://docs.docker.com/build/concepts/context/#dockerignore-files), [named contexts](https://docs.docker.com/build/concepts/context/#named-contexts), other techniques which modify the build context relative to the Dockerfile, e.g. setting the build context explicitly to a subdirectory on the command line:

```bash
docker build -f contrib/Dockerfile src
```

2) When the `COPY` instruction itself is [excluding paths](https://docs.docker.com/reference/dockerfile/#copy---exclude):

```dockerfile
COPY --exclude=.git . .
```

3) In a multi stage build like the following, since the final output image doesn't contain the contents from the build stage with the `COPY . .`:

```dockerfile
FROM golang:1.26 AS build
WORKDIR /src
COPY . .
RUN go build -o myapp .

FROM scratch
COPY --from=build /src/myapp /myapp
ENTRYPOINT /myapp
```
